src/app/interaction-model/InvokeResponder.h

```diff
@@ -103,23 +103,39 @@ class AutoCompleteInvokeResponder
     /// Direct access to reply encoding.
     ///
     /// Use this only in conjunction with the other Raw* calls
-    DataModel::WrappedStructEncoder & RawReplyEncoder(CommandId replyCommandId) { return mWriter->ReplyEncoder(replyCommandId); }
+    DataModel::WrappedStructEncoder & RawResponseEncoder(CommandId replyCommandId)
+    {
+        return mWriter->ResponseEncoder(replyCommandId);
+    }
 
     /// Direct access to flushing replies
     ///
     /// Use this only in conjunction with the other Raw* calls
-    CHIP_ERROR RawFlushPendingReplies() { return mWriter->FlushPendingReplies(); }
+    CHIP_ERROR RawFlushPendingReplies()
+    {
+        mCompleted = false;
+        return mWriter->FlushPendingResponses();
+    }
 
     /// Call "Complete" without the automatic retries.
     ///
     /// Use this in conjunction with the other Raw* calls
-    CHIP_ERROR RawComplete(CHIP_ERROR error) { return mWriter->Complete(error); }
+    CHIP_ERROR RawComplete(CHIP_ERROR error)
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted = true;
```

Comment:
Should we be setting `mCompleted` to true only if `mWriter->Complete(error)` was successful?

Reply:
The intent was to deny the other functions until you call RawFlush. However since this enforcement of logic using a single bool seemed not as robust, I now updated it to a full state machine that enforces the contract of "you may call complete ... and if that fails with buffer too small, you may call flush and complete again once".